website_product_availability/models/product.py

```diff
@@ -50,35 +50,90 @@ class product_product(osv.Model):
         ('4', 'On Request'),
     ]
 
-    def _get_availability(self, cr, uid, ids, name, args, context=None):
-        res = {}
-        for product in self.browse(cr, uid, ids):
-            for route in product.route_ids:
-                if route.consider_on_request:
-                    res[product.id] = '4'
-                    return res
-            if product.qty_available > product.low_stock:
-                res[product.id] = '1'
-                return res
-            elif 0 < product.qty_available <= product.low_stock:
-                res[product.id] = '3'
-                return res
-            elif product.qty_available <= 0:
-                res[product.id] = '2'
-                return res
+    @api.one
+    @api.depends('qty_available', 'low_stock')
+    def _get_availability(self):
+        product = self
+        for route in product.route_ids:
+            if route.consider_on_request:
+                product.stock_state = '4'
+                return
+        if product.qty_available > product.low_stock:
+            product.stock_state = '1'
+        elif 0 < product.qty_available <= product.low_stock:
+            product.stock_state = '3'
+        elif product.qty_available <= 0:
+            product.stock_state = '2'
+
+    low_stock = fields.Integer(
+        'Low Stock',
+        help='This field is used to show '
+        'on the website when a product has '
+        'low availability, any number of stock'
+        'equals the value set or lower will show '
+        '\'low avilability\' on product')
+    stock_state = fields.Selection(
+        compute=_get_availability,
+        selection=SELECTION_LIST,
+        string="Website Stock State",
+        default=0)
+
+
+class LocationQuants(models.Model):
+    _name = 'location.quants'
+
+    product_id = fields.Many2one('product.template', 'Product')
+    location_id = fields.Many2one('stock.location', 'Stock Location')
+    qty = fields.Float('Quantity')
+
+
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    @api.multi
+    def _get_locations_quants(self):
+        stock_locations_obj = self.env['location.quants']
+        stock_quants_obj = self.env['stock.quant']
+        product_id = self.id
+        products = self._get_products()
+        quants = stock_quants_obj.search(
+            [('product_id', 'in', products),
+             ('location_id.usage', '=', 'internal')])
+        stock_locations_ids = []
+        for quant in quants:
+            if quant.location_id.id in stock_locations_ids:
+                continue
+            stock_locations_ids.append(quant.location_id.id)
+        location_quant_ids = stock_locations_obj.search(
+            [('product_id', 'in', products),
+             ('location_id', 'in', stock_locations_ids)])
+        if location_quant_ids:
+            stock_locations_obj.unlink(location_quant_ids)
```

Comment:
You are right, also in that way, method brings a enormous computational cost, I already made the correction, please review.